gitlabform/processors/util/labels_processor.py

```diff
@@ -0,0 +1,97 @@
+from logging import debug, info
+from typing import Dict, List, Callable, Union
+
+from gitlab.base import RESTObject
+from gitlab.v4.objects import Group, Project, ProjectLabel, GroupLabel
+
+
+class LabelsProcessor:
+
+    # Groups and Projects share the same API for .labels within python-gitlab
+    def process_labels(
+        self,
+        configured_labels: Dict,
+        enforce: bool,
+        group_or_project: Union[
+            Group, Project
+        ],  # Group | Project -> |: operand not supported in 3.8/3.9
+        needs_update_method: Callable,
+    ):
+        existing_labels = group_or_project.labels.list()
+        existing_label_names: List = []
+
+        if isinstance(group_or_project, Group):
+            parent_object_type = "Group"
+        else:
+            parent_object_type = "Project"
+
+        if existing_labels:
+            for listed_label in existing_labels:
+                full_label = group_or_project.labels.get(listed_label.id)
+                label_name = full_label.name
+
+                if label_name not in configured_labels.keys():
+                    self.process_label_not_in_config(
+                        enforce, full_label, label_name, parent_object_type
+                    )
+                else:
+                    self.update_existing_label(
+                        configured_labels,
+                        existing_label_names,
+                        full_label,
+                        label_name,
+                        needs_update_method,
+                        parent_object_type,
+                    )
+
+        # add new labels
+        for label_name in configured_labels.keys():
+            if label_name not in existing_label_names:
+                self.create_new_label(
+                    configured_labels, group_or_project, label_name, parent_object_type
+                )
+
+    def process_label_not_in_config(
+        self,
+        enforce: bool,
+        full_label: Union[GroupLabel, ProjectLabel],  # GroupLabel | ProjectLabel
+        label_name: str,
+        parent_object_type: str,
+    ):
+        debug(f"{label_name} not in configured labels")
+        # only delete labels when enforce is true, because user's maybe automatically applying labels based
+        # on Repo state, for example: Compliance Framework labels based on language or CI-template status
+        if enforce:
+            info(f"Removing {label_name} from {parent_object_type}")
+            full_label.delete()
+
+    def update_existing_label(
```

Comment:
The parameters of this function is a bit unusual. We're going to process/update a specific label. It's being called in a loop. Does it need all the labels in the config and all the existing label? Feels like the caller should simply pass the label that should be updated and their current config.

Reply:
@amimas - good points, over-enthusiastic method extraction from me, should be this way round so it's more legible